tests/utils.py

```diff
@@ -21,6 +22,102 @@
+        self.model = model
+        self.params = params
+        self.__class__.__name__ = model.__name__
+
+    def fit(self, X, y):
+        init_params = self.params.get("init", {})
+        self.fit_intercept_ = init_params.pop("fit_intercept", False)
+        if self.fit_intercept_:
+            X = sm.add_constant(X)
+        est = self.model(y, X, **init_params)
+        if "fit_regularized" in self.params:
+            self.fitted_model_ = est.fit_regularized(
+                **self.params["fit_regularized"])
+        else:
+            self.fitted_model_ = est.fit(**self.params.get("fit", {}))
+        self.__class__.__name__ = type(self.fitted_model_).__name__
+        return self.fitted_model_
+
+    def predict(self, X):
+        if self.fit_intercept_:
+            X = sm.add_constant(X)
+        return self.fitted_model_.predict(X)
+
+
+class ModelTrainer:
+
+    _class_instances = {}
+
+    def __init__(self, dataset_name, test_fraction):
+        self.dataset_name = dataset_name
+        self.test_fraction = test_fraction
+        np.random.seed(seed=7)
+        if dataset_name == "boston":
+            self.__name__ = "train_model_regression"
+            dataset = datasets.load_boston()
+            self.X, self.y = shuffle(
+                dataset.data, dataset.target, random_state=13)
+        elif dataset_name == "iris":
+            self.__name__ = "train_model_classification"
+            dataset = datasets.load_iris()
+            self.X, self.y = shuffle(
+                dataset.data, dataset.target, random_state=13)
+        elif dataset_name == "breast_cancer":
+            self.__name__ = "train_model_classification_binary"
+            dataset = datasets.load_breast_cancer()
+            self.X, self.y = shuffle(
+                dataset.data, dataset.target, random_state=13)
+        elif dataset_name == "regression_rnd":
+            self.__name__ = "train_model_regression_random_data"
+            N = 1000
+            self.X = np.random.random(size=(N, 200))
+            self.y = np.random.random(size=(N, 1))
+        elif dataset_name == "classification_rnd":
+            self.__name__ = "train_model_classification_random_data"
+            N = 1000
+            self.X = np.random.random(size=(N, 200))
+            self.y = np.random.randint(3, size=(N,))
+        elif dataset_name == "classification_binary_rnd":
+            self.__name__ = "train_model_classification_binary_random_data"
+            N = 1000
+            self.X = np.random.random(size=(N, 200))
+            self.y = np.random.randint(2, size=(N,))
+        else:
+            raise ValueError("Unknown dataset name: {}".format(dataset_name))
+
+        offset = int(self.X.shape[0] * (1 - test_fraction))
+        self.X_train, self.y_train = self.X[:offset], self.y[:offset]
+        self.X_test, self.y_test = self.X[offset:], self.y[offset:]
+
+    @classmethod
+    def get_instance(cls, dataset_name, test_fraction=0.02):
```

Comment:
I like this caching mechanism 👍

Reply:
Shall we consider making the fitted model a part of the cache as well? Considering that all models are shared between languages. Let me know what you think

Reply:
Thanks!

Seems reasonable to me! I'll try.

Reply:
I've pushed an attempt to cache fitted models in 27ffdd6a588bd0142dd918b71c51d8188a420454. All e2e tests are passed, however, some API tests are failing due to the usage of different params for the same class. The thing is that now I cache **class name** and therefore we cannot differentiate between class instances. Caching **class instances** seems to be pointless as we never use the same instance more than once. Please let me know if you have any workarounds in the mind.

Reply:
To be honest, I'm starting to doubt that caching models is a good idea...

Reply:
> To be honest, I'm starting to doubt that caching models is a good idea...

That's ok, it might as well be not a good idea indeed :)

> The thing is that now I cache class name and therefore we cannot differentiate between class instances. 

Can't you use the estimator's object pointer or some other instance identifier (eg. using `id()`) instead? I thought the point was to prevent fitting the same models more than once in E2E tests. Somehow I thought this approach should help. Am I missing something here?

Reply:
I'd say it's a stretch goal anyway. We can revisit this part later. Shouldn't be a stopper for this PR.

Reply:
> Can't you use the estimator's object pointer or some other instance identifier (eg. using id()) instead? I thought the point was to prevent fitting the same models more than once in E2E tests. Somehow I thought this approach should help. Am I missing something here?

I'm sorry, but I can't get the point of using `id()` (or some unique identifier) as a part of a caching key. Every time we create new estimator object in tests 
https://github.com/BayesWitnesses/m2cgen/blob/d28cb0063ab047751bd72eb7788b109a629f9068/tests/e2e/test_e2e.py#L131
we get new unique key, which results in a cache miss.
```
import lightgbm as lgb
a = lgb.LGBMRegressor()
b = lgb.LGBMRegressor()
id(a) == id(b)
>>> False
```

I tried the following way:
```
        key = dataset_key + " {}".format((hash(type(estimator).__name__
                                          + estimator.__dict__.__repr__()))
                                         if estimator is not None
                                         else "None")
```
But it didn't work for some API tests, because some classes have identical interfaces.

https://travis-ci.org/StrikerRUS/m2cgen/builds/650970719
![image](https://user-images.githubusercontent.com/25141164/74800122-90495400-52e3-11ea-8fd9-47ea9293edc5.png)

Below we can see that for `SVR` model test it uses cashed previously trainer for `NuSVR`.
![image](https://user-images.githubusercontent.com/25141164/74800313-1d8ca880-52e4-11ea-8205-5fbab0ad52ac.png)


If you don't mind, I'm going to revert model caching for now as it's bugging and I'm out of ideas how to fix it. We can get back to it later.

Reply:
> If you don't mind, I'm going to revert model caching for now as it's bugging and I'm out of ideas how to fix it. We can get back to it later.

Yes absolutely. Thank you so much for putting effort into this